Installation issue: lammps

### Steps to reproduce the issue

```console
$ spack spec -I lammps +kokkos +rocm amdgpu_target=gfx90a
Input spec
--------------------------------
 -   lammps+kokkos+rocm amdgpu_target=gfx90a

Concretized
--------------------------------
 -   lammps@20220623.3%gcc@12.2.0~adios~asphere~atc~awpmd~bocs~body~bpm~brownian~cg-dna~cg-sdk~class2~colloid~colvars~compress~coreshell~cuda~cuda_mps~dielectric~diffraction~dipole~dpd-basic~dpd-meso~dpd-react~dpd-smooth~drude~eff~electrode~exceptions~extra-compute~extra-dump~extra-fix~extra-molecule~extra-pair~fep~ffmpeg~granular~h5md~intel~interlayer~ipo~jpeg~kim+kokkos+kspace~latboltz~latte+lib~machdyn~manifold+manybody~mc~meam~mesont~mgpt~misc~ml-hdnnp~ml-iap~ml-rann~ml-snap~mofff+molecule~molfile+mpi~mpiio~netcdf~opencl+openmp~openmp-package~opt~orient~peri~phonon~plugin~plumed~png~poems~ptm~python~qeq~qtb~reaction~reaxff~replica+rigid+rocm~shock~smtbq~sph~spin~srd~tally~uef~voronoi~vtk~yaff amdgpu_target=gfx90a build_system=cmake build_type=RelWithDebInfo fftw_precision=double generator=make lammps_sizes=smallbig arch=linux-sles15-zen3
[^]      ^cmake@3.26.2%gcc@12.2.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-sles15-zen3
[^]          ^ncurses@6.4%gcc@12.2.0~symlinks+termlib abi=none build_system=autotools arch=linux-sles15-zen3
[^]              ^pkgconf@1.8.0%gcc@12.2.0 build_system=autotools arch=linux-sles15-zen3
[^]          ^openssl@1.1.1t%gcc@12.2.0~docs~shared build_system=generic certs=mozilla arch=linux-sles15-zen3
[^]              ^ca-certificates-mozilla@2023-01-10%gcc@12.2.0 build_system=generic arch=linux-sles15-zen3
[^]              ^perl@5.36.0%gcc@12.2.0+cpanm+open+shared+threads build_system=generic arch=linux-sles15-zen3
[^]                  ^berkeley-db@18.1.40%gcc@12.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-sles15-zen3
[^]                  ^bzip2@1.0.8%gcc@12.2.0~debug~pic+shared build_system=generic arch=linux-sles15-zen3
[^]                      ^diffutils@3.9%gcc@12.2.0 build_system=autotools arch=linux-sles15-zen3
[^]                          ^libiconv@1.17%gcc@12.2.0 build_system=autotools libs=shared,static arch=linux-sles15-zen3
[^]                  ^gdbm@1.23%gcc@12.2.0 build_system=autotools arch=linux-sles15-zen3
[^]                      ^readline@8.2%gcc@12.2.0 build_system=autotools patches=bbf97f1 arch=linux-sles15-zen3
[^]              ^zlib@1.2.13%gcc@12.2.0+optimize+pic+shared build_system=makefile arch=linux-sles15-zen3
[^]      ^cray-mpich@8.1.23%gcc@12.2.0+wrappers build_system=generic arch=linux-sles15-zen3
[+]      ^fftw@3.3.10%gcc@12.2.0+mpi~openmp~pfft_patches build_system=autotools precision=double,float arch=linux-sles15-zen3
[^]      ^gmake@4.4.1%gcc@12.2.0~guile build_system=autotools arch=linux-sles15-zen3
[^]      ^hip@5.4.3%gcc@12.2.0~cuda~ipo+rocm build_system=cmake build_type=Release generator=make patches=ca523f1 arch=linux-sles15-zen3
[^]      ^hsa-rocr-dev@5.4.3%gcc@12.2.0+image~ipo+shared build_system=cmake build_type=Release generator=make patches=71e6851 arch=linux-sles15-zen3
[+]      ^kokkos@3.7.01%gcc@12.2.0~aggressive_vectorization~compiler_warnings~cuda~cuda_constexpr~cuda_lambda~cuda_ldg_intrinsic~cuda_relocatable_device_code~cuda_uvm~debug~debug_bounds_check~debug_dualview_modify_check~deprecated_code~examples~explicit_instantiation~hpx~hpx_async_dispatch~hwloc~ipo~memkind~numactl~openmp~openmptarget~pic+profiling~profiling_load_print~pthread~qthread+rocm+serial+shared~sycl~tests~tuning~wrapper amdgpu_target=gfx90a build_system=cmake build_type=RelWithDebInfo generator=make intel_gpu_arch=none std=14 arch=linux-sles15-zen3
[+]      ^llvm-amdgpu@5.4.3%gcc@12.2.0~ipo~link_llvm_dylib~llvm_dylib+openmp+rocm-device-libs build_system=cmake build_type=Release generator=ninja patches=a08bbe1 arch=linux-sles15-zen3
```


### Error message

<details><summary>Error message</summary>

```
==> Installing lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm
==> No binary for lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm found: installing from source
==> Using cached archive: /lustre/orion/csc322/scratch/blue42u/spack-frontier/source/_source-cache/archive/8a/8a276a01b50d37eecfe6eb36f420f354cde51936d20aca7944dea60d3c098c89.tar.gz
==> No patches needed for lammps
==> lammps: Executing phase: 'cmake'
==> lammps: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    '/lustre/orion/csc322/proj-shared/spack-frontier/opt/spack/linux-sles15-zen3/gcc-12.2.0/gmake-4.4.1-wh5zvat6wa4dfym5d6hkrwoao2ufdc3e/bin/make' '-j16'

5 errors found in build log:
     1725    /opt/rocm-5.4.3/bin/hipcc -DFFT_FFTW3 -DFFT_HIPFFT -DKOKKOS_DEPENDENCE -DLAMMPS_GZIP -DLAMMPS_MEMALIGN=64 -DLAMMPS_OMP_COMPAT=4 -DLAMMPS_SMALLBIG -DLMP_KOKKOS -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX -D_
             MPICC_H -Dlammps_EXPORTS -I/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src -I/lustre/orion/csc322/scratch/blue42u/
             spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KSPACE -I/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3n
             evhk4xcuirdso2424zbadoxhnm/spack-src/src/MANYBODY -I/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/MOLECULE -I/lu
             stre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/RIGID -I/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spa
             ck-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KOKKOS -I/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxh
             nm/spack-build-pofp3ne/styles -isystem /opt/cray/pe/mpich/8.1.23/ofi/gnu/9.1/include -isystem /lustre/orion/csc322/scratch/blue42u/spack-frontier/install/linux-sles15-zen3/gcc-12.2.0/fftw-3.3.10-kyxueh
             aly6llmnt2tjw4nrw676q6hcmz/include -isystem /lustre/orion/csc322/scratch/blue42u/spack-frontier/install/linux-sles15-zen3/gcc-12.2.0/kokkos-3.7.01-lwxmrbl5tznutnmhnu45ax4626lxfiy5/include -O2 -g -DNDEB
             UG -std=c++14 -fPIC -march=znver3 -mtune=znver3 -fopenmp=libomp -march=znver2 -mtune=znver2 -fno-gpu-rdc --offload-arch=gfx90a -MD -MT CMakeFiles/lammps.dir/lustre/orion/csc322/scratch/blue42u/spack-fr
             ontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KOKKOS/pair_lj_charmm_coul_charmm_implicit_kokkos.cpp.o -MF CMakeFiles/lammps.dir/lustre/orion/csc322/scratch/b
             lue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KOKKOS/pair_lj_charmm_coul_charmm_implicit_kokkos.cpp.o.d -o CMakeFiles/lammps.dir/lustre/orion/
             csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KOKKOS/pair_lj_charmm_coul_charmm_implicit_kokkos.cpp.o -c /lustre/orion/csc322/
             scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KOKKOS/pair_lj_charmm_coul_charmm_implicit_kokkos.cpp
     1726    [100%] Building CXX object CMakeFiles/lammps.dir/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KOKKOS/region_bloc
             k_kokkos.cpp.o
     1727    /opt/rocm-5.4.3/bin/hipcc -DFFT_FFTW3 -DFFT_HIPFFT -DKOKKOS_DEPENDENCE -DLAMMPS_GZIP -DLAMMPS_MEMALIGN=64 -DLAMMPS_OMP_COMPAT=4 -DLAMMPS_SMALLBIG -DLMP_KOKKOS -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX -D_
             MPICC_H -Dlammps_EXPORTS -I/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src -I/lustre/orion/csc322/scratch/blue42u/
             spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KSPACE -I/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3n
             evhk4xcuirdso2424zbadoxhnm/spack-src/src/MANYBODY -I/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/MOLECULE -I/lu
             stre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/RIGID -I/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spa
             ck-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KOKKOS -I/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxh
             nm/spack-build-pofp3ne/styles -isystem /opt/cray/pe/mpich/8.1.23/ofi/gnu/9.1/include -isystem /lustre/orion/csc322/scratch/blue42u/spack-frontier/install/linux-sles15-zen3/gcc-12.2.0/fftw-3.3.10-kyxueh
             aly6llmnt2tjw4nrw676q6hcmz/include -isystem /lustre/orion/csc322/scratch/blue42u/spack-frontier/install/linux-sles15-zen3/gcc-12.2.0/kokkos-3.7.01-lwxmrbl5tznutnmhnu45ax4626lxfiy5/include -O2 -g -DNDEB
             UG -std=c++14 -fPIC -march=znver3 -mtune=znver3 -fopenmp=libomp -march=znver2 -mtune=znver2 -fno-gpu-rdc --offload-arch=gfx90a -MD -MT CMakeFiles/lammps.dir/lustre/orion/csc322/scratch/blue42u/spack-fr
             ontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KOKKOS/region_block_kokkos.cpp.o -MF CMakeFiles/lammps.dir/lustre/orion/csc322/scratch/blue42u/spack-frontier/b
             uild/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KOKKOS/region_block_kokkos.cpp.o.d -o CMakeFiles/lammps.dir/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/sp
             ack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KOKKOS/region_block_kokkos.cpp.o -c /lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-po
             fp3nevhk4xcuirdso2424zbadoxhnm/spack-src/src/KOKKOS/region_block_kokkos.cpp
     1728    [100%] Linking CXX shared library liblammps.so
     1729    /lustre/orion/csc322/proj-shared/spack-frontier/opt/spack/linux-sles15-zen3/gcc-12.2.0/cmake-3.26.2-2m4skojqcexfmakwahrntemktilxsgz2/bin/cmake -E cmake_link_script CMakeFiles/lammps.dir/link.txt --verb
             ose=1
     1730    /opt/rocm-5.4.3/bin/hipcc -fPIC -O2 -g -DNDEBUG -DKOKKOS_DEPENDENCE -march=znver2 -mtune=znver2 -fno-gpu-rdc -shared -Wl,-soname,liblammps.so.0 -o liblammps.so.0 @CMakeFiles/lammps.dir/objects1.rsp  -W
             l,-rpath,/lustre/orion/csc322/scratch/blue42u/spack-frontier/install/linux-sles15-zen3/gcc-12.2.0/fftw-3.3.10-kyxuehaly6llmnt2tjw4nrw676q6hcmz/lib:/opt/cray/pe/mpich/8.1.23/ofi/gnu/9.1/lib:/opt/rocm-5.
             4.3/llvm/lib:/lustre/orion/csc322/scratch/blue42u/spack-frontier/install/linux-sles15-zen3/gcc-12.2.0/kokkos-3.7.01-lwxmrbl5tznutnmhnu45ax4626lxfiy5/lib64:::::::::::::::::::::::::::::::::::::::::::::::
             :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
             ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: /lustre/orion/csc322/scratch/blue42u/spack-frontier/install/linux-sles15-zen3/gcc-12.2.0/fftw-3.3.10-kyxuehal
             y6llmnt2tjw4nrw676q6hcmz/lib/libfftw3.so -lm -lhipfft -ldl /opt/cray/pe/mpich/8.1.23/ofi/gnu/9.1/lib/libmpi_gnu_91.so /opt/rocm-5.4.3/llvm/lib/libomp.so /lustre/orion/csc322/scratch/blue42u/spack-front
             ier/install/linux-sles15-zen3/gcc-12.2.0/kokkos-3.7.01-lwxmrbl5tznutnmhnu45ax4626lxfiy5/lib64/libkokkoscontainers.so.3.7.01 /lustre/orion/csc322/scratch/blue42u/spack-frontier/install/linux-sles15-zen3
             /gcc-12.2.0/kokkos-3.7.01-lwxmrbl5tznutnmhnu45ax4626lxfiy5/lib64/libkokkoscore.so.3.7.01 -ldl
  >> 1731    ld.lld: error: unable to find library -lhipfft
  >> 1732    clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
  >> 1733    make[2]: *** [CMakeFiles/lammps.dir/build.make:9851: liblammps.so.0] Error 1
     1734    make[2]: Leaving directory '/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-build-pofp3ne'
  >> 1735    make[1]: *** [CMakeFiles/Makefile2:174: CMakeFiles/lammps.dir/all] Error 2
     1736    make[1]: Leaving directory '/lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-build-pofp3ne'
  >> 1737    make: *** [Makefile:139: all] Error 2

See build log for details:
  /lustre/orion/csc322/scratch/blue42u/spack-frontier/build/spack-stage-lammps-20220623.3-pofp3nevhk4xcuirdso2424zbadoxhnm/spack-build-out.txt
```

</details>


### Information on your system

* **Spack:** 0.20.0.dev0 (118d8e4f571a1d22a2922146d4bebc5bb3691182)
* **Python:** 3.6.15
* **Platform:** linux-sles15-zen3
* **Concretizer:** clingo
* OLCF Crusher/Frontier
* The ROCm packages are external: [packages.yaml](https://github.com/spack/spack/files/11174546/packages.yaml.txt)


### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/11180416/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/11180418/spack-build-env.txt)


@rbberger @renjithravindrankannath @srekolam (including maintainers for `hipfft` here)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/lammps/package.py
+++ b/var/spack/repos/builtin/packages/lammps/package.py
@@ -642,6 +642,14 @@ class Lammps(CMakePackage, CudaPackage, ROCmPackage):
     )
     conflicts("+intel", when="%aocc@:3.2.9999", msg="+intel with AOCC requires version 4 or newer")
 
+    # Backport of https://github.com/lammps/lammps/pull/3726
+    conflicts("+kokkos+rocm+kspace", when="@:20210929.3")
+    patch(
+        "https://github.com/lammps/lammps/commit/ebb8eee941e52c98054fdf96ea78ee4d5f606f47.patch?full_index=1",
+        sha256="3dedd807f63a21c543d1036439099f05c6031fd98e7cb1ea7825822fc074106e",
+        when="@20220623.3:20230208 +kokkos +rocm +kspace",
+    )
+
     patch("lib.patch", when="@20170901")
     patch("660.patch", when="@20170922")
     patch("gtest_fix.patch", when="@:20210310 %aocc@3.2.0")

--- a/var/spack/repos/builtin/packages/lammps/package.py
+++ b/var/spack/repos/builtin/packages/lammps/package.py
@@ -650,6 +650,13 @@ class Lammps(CMakePackage, CudaPackage, ROCmPackage):
         when="@20220623.3:20230208 +kokkos +rocm +kspace",
     )
 
+    # Older LAMMPS does not compile with Kokkos 4.x
+    conflicts(
+        "^kokkos @4:",
+        when="@:20221222",
+        msg="LAMMPS is incompatible with Kokkos 4.x until @20230208",
+    )
+
     patch("lib.patch", when="@20170901")
     patch("660.patch", when="@20170922")
     patch("gtest_fix.patch", when="@:20210310 %aocc@3.2.0")

